File: euler11.py
# Каково наибольшее произведение четырех подряд идущих чисел в таблице 20×20,
# расположенных в любом направлении (вверх, вниз, вправо, влево или по
# диагонали)?

from datetime import datetime
from typing import List, Optional

from numpy import array


def readmatrix(
    namefile: Optional[str] = None,
    text: Optional[List[str]] = None,
    mode: Optional[str] = None
) -> List:
    """
    Возвращает матрицу чисел или слов из файла или переданного текста.

    Args:
        namefile: Путь к файлу.
        text: Текст для парсинга.
        mode: Режим чтения ('text' для слов, None для чисел).

    Returns:
        Матрица чисел или список слов.
    """
    matrix = []
    if text is not None:
        for line in range(len(text)):
            matrix.append([int(i) for i in text[line].split(" ")])
    elif mode is None:
        with open(namefile) as f:
            for line in f:
                matrix.append([int(i) for i in line.rstrip('\n').split()])
    elif mode == 'text':
        with open(namefile) as f:
            for line in f:
                matrix = line.replace('"', '').split(',')
    return matrix


def product_multiply(matrix: List[List[int]], n: int) -> List[int]:
    """
    Возвращает список произведений n элементов в строках матрицы.

    Args:
        matrix: Матрица чисел.
        n: Количество элементов для перемножения.

    Returns:
        Список произведений.
    """
    product = []
    for i in range(len(matrix[0]) - n + 1):
        for j in range(len(matrix)):
            num = 1
            for k in range(n):
                num *= matrix[j][i + k]
            product.append(num)
    return product


def product_multiply_diagonals(matrix: List[List[int]], n: int) -> int:
    """
    Перемножает по n элементов матрицы по диагоналям.

    Args:
        matrix: Матрица чисел.
        n: Количество элементов для перемножения.

    Returns:
        Максимальное произведение.
    """
    max_product = 0
    for x in range(len(matrix[0]) - n + 1):
        for y in range(len(matrix[0]) - n + 1):
            # по диагонали вправо вниз
            num = 1
            for i in range(n):
                num *= int(matrix[x + i][y + i])
            if num > max_product:
                max_product = num

            # по диагонали вправо вверх
            num = 1
            for i in range(n):
                num *= int(matrix[x - i][y + i])
            if num > max_product:
                max_product = num

            # по диагонали влево вниз
            num = 1
            for i in range(n):
                num *= int(matrix[x + i][y - i])
            if num > max_product:
                max_product = num

            # по диагонали влево вверх
            num = 1
            for i in range(n):
                num *= int(matrix[x - i][y - i])
            if num > max_product:
                max_product = num
    return max_product


def get_data(url: str, tag: str, n: int) -> str:
    """
    Возвращает чистый текст по n-ному заданному тегу и url.

    Args:
        url: URL-адрес страницы.
        tag: HTML-тег для поиска.
        n: Индекс элемента.

    Returns:
        Текстовое содержимое элемента.
    """
    from bs4 import BeautifulSoup as bs
    from requests import get

    return bs(get(url, timeout=3).text, 'lxml').find_all(tag)[n].text


def euler11() -> None:
    """
    Решение задачи Эйлера №11.

    Каково наибольшее произведение четырех подряд идущих чисел в таблице
    20×20, расположенных в любом направлении?
    """
    n = 4

    # 1 Вариант
    start_time = datetime.now()
    matrix = readmatrix("euler11.txt")
    matrix_transpose = [[0 for _ in range(len(matrix))]
                        for _ in range(len(matrix[0]))]
    for i in range(len(matrix)):
        for j in range(len(matrix[0])):
            matrix_transpose[j][i] = matrix[i][j]

    print(max(
        product_multiply(matrix, n)
        + product_multiply(matrix_transpose, n)
        + [product_multiply_diagonals(matrix, n)]
    ))
    print(datetime.now() - start_time)

    # 2 Вариант
    start_time = datetime.now()
    matrix = array(readmatrix("euler11.txt"))
    matrix_transpose = matrix.transpose()
    print(max(
        product_multiply(matrix, n)
        + product_multiply(matrix_transpose, n)
        + [product_multiply_diagonals(matrix, n)]
    ))
    print(datetime.now() - start_time)

    # 3 Вариант
    start_time = datetime.now()
    url = 'https://euler.jakumo.org/problems/view/11.html'
    matrix = array(readmatrix(text=get_data(url, 'p', 1).strip().split("\n")))
    matrix_transpose = matrix.transpose()
    print(max(
        product_multiply(matrix, n)
        + product_multiply(matrix_transpose, n)
        + [product_multiply_diagonals(matrix, n)]
    ))
    print(datetime.now() - start_time)


if __name__ == "__main__":
    euler11()

File: go/euler11.go
// euler11 находит наибольшее произведение четырех подряд идущих чисел
// в таблице 20×20.
//
// Решение задачи Эйлера №11.
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
	"time"
)

func readMatrix(namefile string) [][]int {
	data, err := os.ReadFile(namefile)
	if err != nil {
		return nil
	}
	var matrix [][]int
	for _, line := range strings.Split(strings.TrimSpace(string(data)), "\n") {
		if line == "" {
			continue
		}
		var row []int
		for _, s := range strings.Fields(line) {
			n, _ := strconv.Atoi(s)
			row = append(row, n)
		}
		matrix = append(matrix, row)
	}
	return matrix
}

func productMultiply(matrix [][]int, n int) []int {
	product := make([]int, 0)
	for i := 0; i <= len(matrix[0])-n; i++ {
		for j := 0; j < len(matrix); j++ {
			num := 1
			for k := 0; k < n; k++ {
				num *= matrix[j][i+k]
			}
			product = append(product, num)
		}
	}
	return product
}

func productMultiplyDiagonals(matrix [][]int, n int) int {
	maxProduct := 0
	for x := 0; x <= len(matrix[0])-n; x++ {
		for y := 0; y <= len(matrix[0])-n; y++ {
			num := 1
			for i := 0; i < n; i++ {
				num *= matrix[x+i][y+i]
			}
			if num > maxProduct {
				maxProduct = num
			}
		}
	}
	return maxProduct
}

func euler11() {
	n := 4

	startTime := time.Now()
	matrix := readMatrix("python/euler11.txt")
	if matrix == nil {
		fmt.Println("File not found")
		return
	}
	matrixTranspose := make([][]int, len(matrix[0]))
	for i := range matrixTranspose {
		matrixTranspose[i] = make([]int, len(matrix))
	}
	for i := 0; i < len(matrix); i++ {
		for j := 0; j < len(matrix[0]); j++ {
			matrixTranspose[j][i] = matrix[i][j]
		}
	}

	maxProd := 0
	for _, p := range productMultiply(matrix, n) {
		if p > maxProd {
			maxProd = p
		}
	}
	for _, p := range productMultiply(matrixTranspose, n) {
		if p > maxProd {
			maxProd = p
		}
	}
	diag := productMultiplyDiagonals(matrix, n)
	if diag > maxProd {
		maxProd = diag
	}
	fmt.Println(maxProd)
	fmt.Println(time.Since(startTime))

	startTime = time.Now()
	maxProd = 0
	for _, p := range productMultiply(matrix, n) {
		if p > maxProd {
			maxProd = p
		}
	}
	for _, p := range productMultiply(matrixTranspose, n) {
		if p > maxProd {
			maxProd = p
		}
	}
	diag = productMultiplyDiagonals(matrix, n)
	if diag > maxProd {
		maxProd = diag
	}
	fmt.Println(maxProd)
	fmt.Println(time.Since(startTime))
}

func main() {
	euler11()
}
